torch/csrc/jit/passes/break_continue_transform.cpp

```diff
@@ -0,0 +1,295 @@
+#include <torch/csrc/jit/passes/break_continue_transform.h>
+#include <ATen/core/jit_type.h>
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/ir_views.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/script/error_report.h>
+#include <torch/csrc/jit/script/final_returns.h>
+
+namespace torch {
+namespace jit {
+namespace script {
+
+void moveBlockBeforeNode(Node* before_node, Block* block);
+
+/**
+ * This pass transforms the Graph so that break & continue statements are
+ * removed. We transform the graph so that ops following a break or continue are
+ * not run.
+ */
+
+// Will a block or node continue or break
+enum LoopStatus { WONT, MIGHT, WILL };
+
+// Are we transforming breaks or continues
+enum Transform { BREAKS, CONTINUES };
+
+struct LoopTransformer {
+  LoopTransformer(std::shared_ptr<Graph> graph_, Transform transform_)
+      : graph(std::move(graph_)) {
+    WithInsertPoint guard(graph->block()->nodes().front());
+    true_val = graph->insertConstant(true);
+    false_val = graph->insertConstant(false);
+    transform = transform_;
+  };
+
+  const std::string& getVarname() {
```

Comment:
This isn't going to work for nested break/continue right? All the guards have the same names.

Reply:
We don't execute any additional ops after a break has been hit, so there is no way to set break to false after it has been true. I think I just need to set $did_break to false at the start of a loop to make sure that it does not accidentally get the value from a previous loop.